### ISSUE TITLE
changing items (hence list size) when fully scrolled causes problems

### ISSUE THREAD

**Issue by m4b** [PolymerElements/iron-list#12]
Very briefly, I've noticed that:

1. If the user scrolls all the way down on a list,
2. Then effects a new dataset/array change, which has a smaller amount of elements (and hence less height w.r.t the scroller)
3. Then the list behaves "unusually"; in Chrome the scroller doesn't appear to move when the user clicks on it and they have to "scroll" all the way to the top for the list to show items again.

I can mock up a simple example if the above isn't sufficient

**Comment by blasten** [comment]
@m4b I tested that case. for example:  scroll all the way down to the bottom of the list or somewhere in the middle and then change the items:

`list.items = list.items.splice(0, 20);`
or
`list.splice('items', 0, 20)`

**Comment by m4b** [comment]
@blasten and the item list is smaller?  Let me cook up a simple example trying to reproduce then, maybe something else is causing it in the larger app that I'm seeing it in.  brb ;)

**Comment by blasten** [comment]
@m4m yes. In the examples, the list has 500 items

**Comment by m4b** [comment]
@blasten: so I'm able to reproduce it in a somewhat convoluted example with a nested header panel and a `fit` class applied to the list (and in other variations a div containing the list).

It seems to be related to using fit in these cases, but the smaller item array switched out has to still be big enough to cause a scroller to appear.  

When I replace the list with a `<template is="dom-repeat">` it fits as expected, but that doesn't mean much.  I might just not being using it properly?

Here is a jsbin that doesn't work yet, because `<iron-list>` isn't on http://element-party.xyz (yet).  You can add an iron-list import to the bin if you know of a CDN with one, or just copy paste below and reference the iron list from your local bower_components.

http://jsbin.com/wuyomotora/edit?html,console,output

````html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
   <base href="http://element-party.xyz/">
  <script src="bower_components/webcomponentsjs/webcomponents-min.js"></script>
    <!-- iron-list is missing from element-party, need to update -->
  <link rel="import" href="all-elements.html">
</head>
<body>

  <base-panel></base-panel>

  <dom-module id="base-panel">
    <style>
      neon-animated-pages {
        position: static;
      }
    </style>
    <template>
      <paper-drawer-panel>
        <paper-header-panel drawer>
          <paper-toolbar>
            <div>Drawer</div>
          </paper-toolbar>
          <paper-menu selected="{{_selected}}">
            <paper-item>Item 1</paper-item>
            <paper-item>Item 2</paper-item>
          </paper-menu>
        </paper-header-panel>
        <paper-header-panel mode="standard" main>
          <paper-toolbar>
            <h1>[[_selected]]</h1>
          </paper-toolbar>
          <neon-animated-pages selected="{{_selected}}">
            <nested-panel></nested-panel>
            <div>Div</div>
          </neon-animated-pages>
        </paper-header-panel>
      </paper-drawer-panel>
    </template>
    <script>
      Polymer({
        is: "base-panel",
        properties: {
          _selected: {
            type: Number,
            value: 0,
          }
        }
      });
    </script>
  </dom-module>

  <dom-module id="nested-panel">
    <style>
      @apply(--layout-fit);
      paper-material {
        margin: 16px;
      }

      iron-list {}
    </style>
    <template>

      <iron-pages selected="{{_selected}}">
        <paper-material elevation="4" class="fit">
          <paper-header-panel mode="standard">
            <paper-toolbar>
              <paper-button raised icon="refresh" style="color: #00ff00;" on-tap="_changeItems">
                Tap Me To Change Items
              </paper-button>
              <div>Nested Toolbar</div>
            </paper-toolbar>
            <iron-list items="[[_items]]" as="item">
              <template>
                <paper-item>[[item]]</paper-item>
              </template>
            </iron-list>
          </paper-header-panel>
        </paper-material>
        <paper-material>
          <div>I'm unused</div>
        </paper-material>
      </iron-pages>

    </template>
    <script>
      Polymer({
        is: "nested-panel",
        behaviors: [
          Polymer.NeonAnimatableBehavior
        ],
        properties: {

          _current: {
            value: 0
          },

          entryAnimation: {
            value: 'scale-up-animation'
          },
          exitAnimation: {
            value: 'fade-out-animation'
          },
          _selected: {
            type: Number,
            value: 0,
            notify: true
          },
          _items: {
            type: Array,
            value: function() {
              return ['I', 'have', 'of', 'late,', '(but', 'wherefore', 'I', 'know', 'not)', 'lost', 'all', 'my', 'mirth,', 'forgone', 'all', 'custom', 'of', 'exercises;', 'and', 'indeed,',
                'it', 'goes', 'so', 'heavily', 'with', 'my', 'disposition;', 'that', 'this', 'goodly', 'frame', 'the', 'earth,', 'seems', 'to', 'me', 'a', 'sterile', 'promontory;', 'this',
                'most', 'excellent', 'canopy', 'the', 'air,', 'look', 'you,', 'this', 'brave', "o'er", 'hanging', 'firmament,', 'this', 'majestical', 'roof,', 'fretted', 'with', 'golden', 'fire:', 'why,',
                'it', 'appeareth', 'no', 'other', 'thing', 'to', 'me,', 'than', 'a', 'foul', 'and', 'pestilent', 'congregation', 'of', 'vapours.', "'What", 'a', 'piece', 'of', 'work',
                'is', 'a', 'man!', 'How', 'noble', 'in', 'reason,', 'how', 'infinite', 'in', 'faculty!', 'In', 'form', 'and', 'moving', 'how', 'express', 'and', 'admirable!', 'In',
                'action', 'how', 'like', 'an', 'Angel!', 'in', 'apprehension', 'how', 'like', 'a', 'god!', 'The', 'beauty', 'of', 'the', 'world!', 'The', 'paragon', 'of', 'animals!',
                'And', 'yet', 'to', 'me,', 'what', 'is', 'this', 'quintessence', 'of', 'dust?', 'Man', 'delights', 'not', 'me;', 'no,', 'nor', 'Woman', 'neither;', 'though', 'by',
                'your', 'smiling', 'you', 'seem', 'to', 'say', 'so.'
              ];
            }
          }
        },

        _changeItems: function() {

          this._current = (this._current + 1) % 2;
          if (this._current % 2 === 0) {
            this._items = ['testing', 'this', 'smaller', 'array', 'how', 'infinite', 'in', 'faculty!', 'In', 'form', 'and', 'moving', 'how', 'express', 'and', 'admirable!', 'In',
              'action', 'how', 'like', 'an', 'Angel!', 'in', 'apprehension', 'how', 'like', 'a', 'god!', 'The', 'beauty', 'of', 'the', 'world!', 'The', 'paragon', 'of', 'animals!',
              'And', 'yet', 'to', 'me,', 'what', 'is', 'this', 'quintessence', 'of', 'dust?', 'Man', 'delights', 'not', 'me;', 'no,', 'nor', 'Woman', 'neither;', 'though', 'by',
              'your', 'smiling', 'you', 'seem', 'to', 'say', 'END OF ELEMENTS'
            ];
          } else {
            this._items = ['I', 'have', 'of', 'late,', '(but', 'wherefore', 'I', 'know', 'not)', 'lost', 'all', 'my', 'mirth,', 'forgone', 'all', 'custom', 'of', 'exercises;', 'and', 'indeed,',
              'it', 'goes', 'so', 'heavily', 'with', 'my', 'disposition;', 'that', 'this', 'goodly', 'frame', 'the', 'earth,', 'seems', 'to', 'me', 'a', 'sterile', 'promontory;', 'this',
              'most', 'excellent', 'canopy', 'the', 'air,', 'look', 'you,', 'this', 'brave', "o'er", 'hanging', 'firmament,', 'this', 'majestical', 'roof,', 'fretted', 'with', 'golden', 'fire:', 'why,',
              'it', 'appeareth', 'no', 'other', 'thing', 'to', 'me,', 'than', 'a', 'foul', 'and', 'pestilent', 'congregation', 'of', 'vapours.', "'What", 'a', 'piece', 'of', 'work',
              'is', 'a', 'man!', 'How', 'noble', 'in', 'reason,', 'how', 'infinite', 'in', 'faculty!', 'In', 'form', 'and', 'moving', 'how', 'express', 'and', 'admirable!', 'In',
              'action', 'how', 'like', 'an', 'Angel!', 'in', 'apprehension', 'how', 'like', 'a', 'god!', 'The', 'beauty', 'of', 'the', 'world!', 'The', 'paragon', 'of', 'animals!',
              'And', 'yet', 'to', 'me,', 'what', 'is', 'this', 'quintessence', 'of', 'dust?', 'Man', 'delights', 'not', 'me;', 'no,', 'nor', 'Woman', 'neither;', 'though', 'by',
              'your', 'smiling', 'you', 'seem', 'to', 'say', 'so.'
            ];
          }
          console.log("changing: ", this._current, this._items);
        }
      });
    </script>
  </dom-module>

</body>
</html>
````

**Comment by m4b** [comment]
@blasten: here is a simpler example, which breaks the scroller on chrome when you change the items when scrolled all the way to the bottom.  In some cases (in a simple psk project structure, for example), the items all appear on the same line initially on chrome.
 
On firefox the scroller doesn't break when changing elements when scrolled all the way to the bottom, but after scrolling past the first page of elements, the items no longer appear.

On safari the breakage is similar to chrome.

JSBIN: http://jsbin.com/fiheji/edit?html,console,output

Module:
````html
<dom-module id="list-fit">
  <style>
  @apply(--layout-fit);
  paper-material {
    margin: 16px;
  }
  </style>
  <template>
      <paper-material elevation="4" class="fit">
        <paper-header-panel mode="standard">
          <paper-toolbar>
            <paper-button raised on-tap="changeItems">Change Items</paper-button>
            <div>Iron list inside fit</div>
          </paper-toolbar>
            <iron-list items="[[_items]]">
              <template>
                <paper-item>[[item]]</paper-item>
              </template>
            </iron-list>
        </paper-header-panel>
      </paper-material>
  </template>
  <script>
  Polymer({
    is: 'list-fit',
    properties: {
      _current: {
        value: 0
      },
      _items: {
        type: Array,
        notify: true,
        value: function() {
          var i = 500;
          var res = [];
          while (i--){
            res.push(i);
          }
          return res;
        }
      }
    },
    changeItems: function(){
      this._current = (this._current +1) % 2;
      var i = 500;
      if (this._current % 2 === 0){
        i = 500;
      }else{
        i = 100;
      }
      var tmp = [];
      while (i--){
         tmp.push(i);
      }
      this._items = tmp;
      console.log('item count: ', this._items.length);
    }
  });
</script>
</dom-module>
````

I've also noticed that embedding `<iron-list>` inside nested `<header-panel>` with `class="fit"` on a parent element also cause similar breakage, with elements not showing up, etc.

This seems like fairly common behavior to have the main UI element fit, especially when nested in other elements, but please correct me if I'm wrong here, what's recommended instead if so, etc.

**Comment by blasten** [comment]
@m4b This issue should be fixed by now. Could you give it a try?

**Comment by m4b** [comment]
@blasten: just tested the original jsbin (the server `bower update`s the elements every day), http://jsbin.com/fiheji/edit?html,console,output, and everything seems to be working great, nice work!

I also noticed the scroll-to-top is working correctly; when changing items, it correctly pops to the top of the list, and the items display correctly (without any momentary blank).

Unfortunately, when I tested on a more complicated app, on Firefox, Chrome, and android, the issue still seems to persist w.r.t the scroll top not being reset to zero after items are updated.  I'm not sure why, as the app example is very similar in nature to the jsbin, except it _does_ employ a nested `paper-header-panel`.

I'm a little busy now, but I hope to submit another issue with an example that attempts to reproduce it  soon (the scroll top not being reset).

But, since this issue is fixed, I'm going to close, again, nice work, I think we're almost there!
